File: Python/Subsets.py
# Given a set of distinct integers, S, return all possible subsets.

# Note:
# Elements in a subset must be in non-descending order.
# The solution set must not contain duplicate subsets.
# For example,
# If S = [1,2,3], a solution is:

# [
  # [3],
  # [1],
  # [2],
  # [1,2,3],
  # [1,3],
  # [2,3],
  # [1,2],
  # []
# ]


class Solution:
    # @param S, a list of integer
    # @return a list of lists of integer
	def subsets(self, S):
		result = []
		output = []
		if len(S) == 0:
			return result
		result.append(output)
		S.sort()
		self.generateSub(S,0,result,output)
		return result
	def generateSub(self,S,step,result,output):
		for i in range(step,len(S)):
			output.append(S[i])
			result.append(list(output))
			if i < (len(S) - 1):
				self.generateSub(S,i+1,result,output)
			output.pop()

File: Java/Subsets.java
class Subsets{
	public ArrayList<ArrayList<Integer>> subsets(int[] S) {
		ArrayList<ArrayList<Integer>> res = new ArrayList<ArrayList<Integer>>();
        ArrayList<Integer> solution = new ArrayList<Integer>();
        if(S == null || S.length == 0)
            return res;
        Arrays.sort(S);
        recursive(S,res,solution,0);
        res.add(new ArrayList<Integer>());
        return res;
	}
	private void recursive(int[] S, ArrayList<ArrayList<Integer>> res, 
			ArrayList<Integer> solution, int level){
		if(level == S.length){
			return;
		}
		for(int i=level;i<S.length;i++){
			solution.add(S[i]);
			res.add(new ArrayList<Integer>(solution));
			recursive(S,res,solution,i+1);
			solution.remove(solution.size()-1);
		}
	}
	private ArrayList<ArrayList<Integer>> iterativeSolution(int[] S){
		ArrayList<ArrayList<Integer>> res = new ArrayList<ArrayList<Integer>>();
		for(int i=0;i<S.length;i++){
			ArrayList<ArrayList<Integer>> temp = new ArrayList<ArrayList<Integer>>();
			for(ArrayList<Integer> al:res)
				temp.add(new ArrayList<Integer>(al));
			for(ArrayList<Integer> al:temp){
				al.add(S[i]);
			}
			ArrayList<Integer> a = new ArrayList<Integer>();
			a.add(S[i]);
			res.add(a);
			res.addAll(temp);
		}
		res.add(new ArrayList<Integer>());
        return res;
	}
}